### DISCUSSION TITLE
An error occurred (NoSuchUpload) when calling the CompleteMultipartUpload operation

### DISCUSSION THREAD

**Discussion by richardARPANET** [minio/minio#17420] [Q&A]
Version RELEASE.2023-06-02T23-17-26Z
This issue only happens once the app is deployed, it does not happen when the application is ran on local machine (even with the same test upload files).

Everything also works fine with smaller file uploads (500MB), this happens with a 1GB file. 
File parts are uploaded with chunk size of 5MB.

Here is what is happening:

1. Init multipart upload (server side, return presigned URLS to JS frontend)
2. Upload parts (in web browser [JS code here](https://github.com/VeemsHQ/veems/blob/develop/react-components/src/api/api.js#L332))
3. Call 'complete' (server side, [Python code here](https://github.com/VeemsHQ/veems/blob/5c76b8afd3fb221e35541d1a66ba7f7018b4725b/veems/media/upload_manager.py#L135))
4. Minio complains the upload doesn't exist

```
NoSuchUpload

An error occurred (NoSuchUpload) when calling the CompleteMultipartUpload operation: The specified multipart upload does not exist. The upload ID may be invalid, or the upload may have been aborted or completed.
```

![image](https://github.com/minio/minio/assets/38049750/650c0c99-4389-423c-878b-10fd31a400d0)

Why would this be happening and how to get more information from the server as to why that uploadId magically no longer exists?

**Comment by klauspost** [comment]
1) Post (ascii, not screenshots) of all requests. If possible also response bodies.
2) Capture `mc admin trace -v ALIAS` for the same requests.

I suspect a trivial mistake, like a different bucket. Obviously we have many tests in place for this basic functionality.

**Reply by richardARPANET** [reply]
Thanks @klauspost, doing this now. Will report back

**Reply by richardARPANET** [reply]
This is still happening and seemingly randomly, so I cannot reliably run "trace" while it happens. Whenever I try, it never happens :shrug:

**Reply by richardARPANET** [reply]
@klauspost this just occurred while I was tracing:

trace: https://gist.github.com/richardARPANET/4bdaa68dc3088c6e9061c3ed2bd2eb98
error:

```
NoSuchUpload

An error occurred (NoSuchUpload) when calling the CompleteMultipartUpload operation: The specified multipart upload does not exist. The upload ID may be invalid, or the upload may have been aborted or completed.
```

some context for the failed request from the Sentry error page:

```
api_params	

{Bucket: 'example-media', Key: 'uploads/MWTWN5IvQqa2.mp4', MultipartUpload: {Parts: [{"ETag":"'\"2f4b3e414cdfb938ad91c239ee8dcb40\"'","PartNumber":"1"}, {"ETag":"'\"53694dc5fa0ef789e867bab54b819e4b\"'","PartNumber":"2"}, {"ETag":"'\"827a4a53bfbaedf96a21055df9fe8009\"'","PartNumber":"3"}, {"ETag":"'\"09af71491c89c5de0b18663a7d024c22\"'","PartNumber":"4"}, {"ETag":"'\"c6e8e847e07689afab490d6ddfa760ad\"'","PartNumber":"5"}, {"ETag":"'\"5e85131e312f36ab9b732786e4d2ffe4\"'","PartNumber":"6"}, {"ETag":"'\"ca059990995df9c2422b92766c1c7ea8\"'","PartNumber":"7"}, {"ETag":"'\"2b519d06471cb9e04ab741d463627bd6\"'","PartNumber":"8"}, {"ETag":"'\"04b38c5ad50a94651465e6e510b85c11\"'","PartNumber":"9"}, {"ETag":"'\"cf75ed12b66a8e0e768e8c6779ca9d04\"'","PartNumber":"10"}]}, UploadId: 'NjQ0NzJkMWEtNzQzMC00OTM5LWI5NzMtMmM4NTY2MGRlMGI0LjkyYjAwNjVlLWVhOGEtNGFhNS04MDM0LWUwZTk4NzA3Mjk3NA'}

event_response	

None

handler	

<function inject_api_version_header_if_needed at 0x7fc43504d1b0>

operation_model	

OperationModel(name=CompleteMultipartUpload)

operation_name	

'CompleteMultipartUpload'

request_context	

{auth_type: [Filtered], client_config: <botocore.config.Config object at 0x7fc4185e1600>, client_region: 'us-east-1', has_streaming_input: False, retries: {attempt: 1, invocation-id: 'f20de0d2-140b-462c-be76-3377fb9a1f45', max: 5}, signing: {bucket: 'example-media'}, timestamp: '20230619T131221Z'}

request_dict	

{body: b'<CompleteMultipartUpload xmlns="http://s3.amazonaws.com/doc/2006-03-01/"><Part><ETag>"2f4b3e414cdfb938ad91c239ee8dcb40"</ETag><PartNumber>1</PartNumber></Part><Part><ETag>"53694dc5fa0ef789e867bab54b819e4b"</ETag><PartNumber>2</PartNumber></Part><Part><ETag>"827a4a53bfbaedf96a21055df9fe8009"</ETag><PartNumber>3</PartNumber></Part><Part><ETag>"09af71491c89c5de0b18663a7d024c22"</ETag><PartNumber>4</PartNumber></Part><Part><ETag>"c6e8e847e07689afab490d6ddfa760ad"</ETag><PartNumber>5</PartNumber></Part><Par..., context: {auth_type: [Filtered], client_config: <botocore.config.Config object at 0x7fc4185e1600>, client_region: 'us-east-1', has_streaming_input: False, retries: {

attempt: 1, invocation-id: 'f20de0d2-140b-462c-be76-3377fb9a1f45', max: 5}, signing: {

bucket: 'example-media'}, timestamp: '20230619T131221Z'}, headers: {User-Agent: 'Boto3/1.21.38 Python/3.10.12 Linux/5.15.0-73-generic Botocore/1.24.38'}, method: 'POST', query_string: {uploadId: 'NjQ0NzJkMWEtNzQzMC00OTM5LWI5NzMtMmM4NTY2MGRlMGI0LjkyYjAwNjVlLWVhOGEtNGFhNS04MDM0LW...}}
```

**Reply by klauspost** [reply]
It seems like you have multiple calls to the `CompleteMultipartUpload`:

```
minio-tenant-dev-pool-0-0.minio-tenant-dev-hl.minio-dev.svc.cluster.local:9000 [REQUEST s3.CompleteMultipartUpload] [2023-06-19T13:00:26.745] [Client IP: 10.42.2.0]
minio-tenant-dev-pool-0-0.minio-tenant-dev-hl.minio-dev.svc.cluster.local:9000 POST /example-media/uploads/MWTWN5IvQqa2.mp4?uploadId=NjQ0NzJkMWEtNzQzMC00OTM5LWI5NzMtMmM4NTY2MGRlMGI0LjkyYjAwNjVlLWVhOGEtNGFhNS04MDM0LWUwZTk4NzA3Mjk3NA
...
minio-tenant-dev-pool-0-0.minio-tenant-dev-hl.minio-dev.svc.cluster.local:9000 X-Amz-Date: 20230619T130026Z
minio-tenant-dev-pool-0-0.minio-tenant-dev-hl.minio-dev.svc.cluster.local:9000 <CompleteMultipartUpload xmlns="http://s3.amazonaws.com/doc/2006-03-01/"><Part><ETag>"2f4b3e414cdfb938ad91c239ee8dcb40"</ETag><PartNumber>1</PartNumber>....</CompleteMultipartUpload>
minio-tenant-dev-pool-0-0.minio-tenant-dev-hl.minio-dev.svc.cluster.local:9000 [RESPONSE] [2023-06-19T13:00:27.194] [ Duration 449.553ms  ↑ 4.6 KiB  ↓ 360 B ]
minio-tenant-dev-pool-0-0.minio-tenant-dev-hl.minio-dev.svc.cluster.local:9000 200 OK
```

```
minio-tenant-dev-pool-0-0.minio-tenant-dev-hl.minio-dev.svc.cluster.local:9000 [REQUEST s3.CompleteMultipartUpload] [2023-06-19T13:12:21.185] [Client IP: 10.42.2.0]
minio-tenant-dev-pool-0-0.minio-tenant-dev-hl.minio-dev.svc.cluster.local:9000 POST /example-media/uploads/MWTWN5IvQqa2.mp4?uploadId=NjQ0NzJkMWEtNzQzMC00OTM5LWI5NzMtMmM4NTY2MGRlMGI0LjkyYjAwNjVlLWVhOGEtNGFhNS04MDM0LWUwZTk4NzA3Mjk3NA
...
minio-tenant-dev-pool-0-0.minio-tenant-dev-hl.minio-dev.svc.cluster.local:9000 <CompleteMultipartUpload xmlns="http://s3.amazonaws.com/doc/2006-03-01/"><Part><ETag>"2f4b3e414cdfb938ad91c239ee8dcb40"</ETag><PartNumber>1</PartNumber>...</CompleteMultipartUpload>
minio-tenant-dev-pool-0-0.minio-tenant-dev-hl.minio-dev.svc.cluster.local:9000 [RESPONSE] [2023-06-19T13:12:21.200] [ Duration 14.123ms  ↑ 4.6 KiB  ↓ 473 B ]
minio-tenant-dev-pool-0-0.minio-tenant-dev-hl.minio-dev.svc.cluster.local:9000 404 Not Found
```

Once the multipart upload has been converted to an object, it no longer "exists". The S3 spec does not guarantee that these calls are idempotent for any specific amount of time. From [here](https://docs.aws.amazon.com/amazonglacier/latest/dev/api-multipart-complete-upload.html):

> Complete Multipart Upload is an idempotent operation. After your first successful complete multipart upload, if you call the operation again within a short period, the operation will succeed and return the same archive ID.

The 12 minutes between the calls means the upload-id no longer exists. I am not sure if there is a fixed time limit, so you should avoid relying on this functionality.

**Reply by richardARPANET** [reply]
@klauspost I just noticed this. I think it's because this "complete" call is ran in a background job.

1. Client makes HTTP call to "complete" API.
2. That API triggers background job that does the multipart complete.

If that background job fails, it will auto-retry (with some back-off delay). So perhaps that happens?

So I'm wondering is there a way to check the status of a multipart upload? That way perhaps in this background job it can check the status to avoid multiple calls to 'complete'? Which is what could be causing the error.

**Reply by klauspost** [reply]
@richardARPANET It seems to be something internal in Boto3 that does this. I don't know what can cause it to delay a second call by 12 minutes.
